packages/api/src/middleware/validation/checkAnimalOrBatch.js

```diff
@@ -13,81 +13,609 @@
+      { error: 'Invalid ids', invalidIds },
+    );
+  }
+};
+
+// For edit mode set creating to false
+const checkAnimalType = async (animalOrBatch, farm_id, creating = true) => {
+  const { default_type_id, custom_type_id, type_name } = animalOrBatch;
+  const typeKeyOptions = ['default_type_id', 'custom_type_id', 'type_name'];
+  // Skip if all undefined or editing (!creating)
+  if (creating || oneExists(typeKeyOptions, animalOrBatch)) {
+    checkExactlyOneIsProvided(
+      [default_type_id, custom_type_id, type_name],
+      'default_type_id, custom_type_id, or type_name',
+    );
+    if (!creating) {
+      // Overwrite with null in db if editing
+      setFalsyValuesToNull(typeKeyOptions, animalOrBatch);
+    }
+    if (custom_type_id) {
+      checkIdIsNumber(custom_type_id);
+      const customType = await CustomAnimalTypeModel.query().findById(custom_type_id);
+      if (!customType) {
+        throw customError('Custom type does not exist');
+      }
+      await checkRecordBelongsToFarm(customType, farm_id, 'custom type');
+    }
+  }
+};
+
+const checkDefaultBreedMatchesType = async (
+  animalOrBatchRecord,
+  default_breed_id,
+  default_type_id,
+) => {
+  // One of these two should exist at this point
+  let defaultBreedId = default_breed_id;
+  let defaultTypeId = default_type_id;
+
+  // If breed or type is not changed get from record
+  if (!defaultBreedId && animalOrBatchRecord) {
+    defaultBreedId = animalOrBatchRecord.default_breed_id;
+  }
+  if (!defaultTypeId && animalOrBatchRecord) {
+    defaultTypeId = animalOrBatchRecord.default_type_id;
+  }
+
+  if (defaultTypeId && defaultBreedId) {
+    checkIdIsNumber(defaultBreedId);
+    const defaultBreed = await DefaultAnimalBreedModel.query().findById(defaultBreedId);
+    if (!defaultBreed) {
+      throw customError('Default breed does not exist');
+    }
+    if (defaultBreed.default_type_id !== defaultTypeId) {
+      throw customError('Breed does not match type');
+    }
+  } else if (!defaultTypeId) {
+    throw customError('Default breed must use default type');
+  }
+};
+
+const checkCustomBreedMatchesType = (
+  animalOrBatch,
+  animalOrBatchRecord,
+  customBreed,
+  default_type_id,
+  custom_type_id,
+) => {
+  // customBreed exists at this point
+  let defaultTypeId = default_type_id;
+  let customTypeId = custom_type_id;
+  const typeKeyOptions = ['default_type_id', 'custom_type_id', 'type_name'];
+
+  // If not editing type, check record type
+  if (!oneExists(typeKeyOptions, animalOrBatch) && animalOrBatchRecord) {
+    defaultTypeId = animalOrBatchRecord.default_type_id;
+    customTypeId = animalOrBatchRecord.custom_type_id;
+  }
+
+  // Custom breed does not match type if defaultId OR customTypeId does not match
+  if (
+    (customBreed.default_type_id && customBreed.default_type_id !== defaultTypeId) ||
+    (customBreed.custom_type_id && customBreed.custom_type_id !== customTypeId)
+  ) {
+    throw customError('Breed does not match type');
```

Comment:
I'm getting this error when I change type to an existing custom type and either:
- send a new breed
- don't send a breed at all

I didn't look too much into the code, but maybe an allowance for `breed_name` needs to go into this section to handle the first case?

And as for not sending a breed... it's not that the breed was unchanged, it's that it was not specified on the new type. Following the POST format, that means the request contained no breed property. Does the API expect a `null` breed for that situation? Or is omitting the breed correct?

Reply:
Yeah I looked into that but didn't want to change the existing order of things. I will look into those cases -- so the two edit cases not working are: 

``` 
{custom_type_id: existingType.id, breed_name: 'name'}
```
And
```
{custom_type_id: existingType.id}
```

Reply:
@Duncan-Brain the second one was this:

```
POST
{... custom_type_id: someType.id (breed or new breed)}
PATCH
{custom_type_id: someOtherType.id -- no breed specified--}
```

Reply:
@kathyavini -- Okay I think this is fixed, I have some tests.

For the second case where someOtherType.id may be correctly failing if someOtherType does not match the original breed. But it should accept a null value for the previously supplied value. I was failing there but should now be fixed.